torch/nn/parameter.py

```diff
@@ -42,3 +46,82 @@ def __reduce_ex__(self, proto):
             torch._utils._rebuild_parameter,
             (self.data, self.requires_grad, OrderedDict())
         )
+
+
+class UninitializedParameter(Parameter):
+    r"""A parameter that is not yet initialized for shape inference support.
+
+    Uninitialized Parameters holds empty tensors that can be moved
+    to different devices and change their type so when they are materialized,
+    they do it directly with the right configuration.
+    """
+    def __new__(cls, requires_grad=True):
+        data = torch.Tensor()
+        return torch.Tensor._make_subclass(cls, data, requires_grad)
+
+    def materialize(self, shape, device=None, dtype=None):
+        r"""Create a Parameter with the same properties of the uninitialized one.
+
+        Given a shape, it materializes a parameter in the same device
+        and with the same `dtype` as the current one or the specified ones in the 
+        arguments
+
+        Args:
+            shape : (tuple): the shape for the materialized tensor.
+            device (:class:`torch.device`): the desired device of the parameters
+                and buffers in this module. Optional.
+            dtype (:class:`torch.dtype`): the desired floating point type of
+                the floating point parameters and buffers in this module. Optional.
+        """
+        if device is None:
+            device = self.data.device
+        if dtype is None:
+            dtype = self.data.dtype
+        return Parameter(torch.empty(shape, device=device, dtype=dtype), requires_grad=self.requires_grad)
+
+    def share_memory_(self):
+        raise RuntimeError(
+            'Can\'t share memory on an uninitialized parameter. '
+            'Run forward to initialize the network before calling '
+            '`module.share_memory()`.')
+
+    def __repr__(self):
+        return 'Uninitialized parameter'
+
+
+class UninitializedBuffer(torch.Tensor):
```

Comment:
You can add this next to Parameter as well.
The base Buffer class will be added there as well when we add it.